common/tools/dev-tool/src/framework/check.ts

```diff
@@ -0,0 +1,297 @@
+   *
+   * If no tags are specified the check will always be run.
+   */
+  tags?: CheckTags[];
+
+  /**
+   * Severity of the check. Defaults to "error".
+   */
+  severity?: CheckSeverity;
+
+  /**
+   * Optional function to determine whether this check applies to the given project
+   * If this function returns true for the project, the check will be run.
+   * If the function returns false, the check will be skipped for the project.
+   *
+   * If left undefined, the check will be enabled for all projects.
+   */
+  enable?: CheckEnableFunction;
+}
+
+export interface CheckContext {
+  /**
+   * Whether to attempt to fix
+   */
+  fix: boolean;
+
+  /**
+   * Information about the project
+   */
+  project: ProjectInfo;
+
+  /**
+   * Whether to output verbosely
+   */
+  verbose: boolean;
+}
+
+export type CheckFunction = (ctx: CheckContext) => Promise<void>;
+
+export interface Check extends CheckOptions {
+  /**
+   * Run the check in the given context
+   */
+  check: CheckFunction;
+}
+
+/**
+ * Error indicating that a check has failed
+ */
+export class CheckFailedError extends Error {
+  constructor(
+    message: string,
+    public detail?: string,
+  ) {
+    super(message);
+    this.name = "CheckFailedError";
+  }
+}
+
+export function isCheckFailedError(e: unknown): e is CheckFailedError {
+  return (e as any)?.name === "CheckFailedError";
+}
+
+/**
+ * Check that condition is true.
+ * @param condition - condition to check
+ * @param message - short message to display
+ * @param detail - more details to display about the failure; can be multiple lines
+ */
+export function check(condition: boolean, message: string, detail?: string): void {
+  if (!condition) {
+    throw new CheckFailedError(message, detail);
+  }
+}
+
+export interface ScriptCheckOptions extends Omit<CheckOptions, "hasFix"> {
+  /**
+   * Command to be ran in order to run the check.
+   *
+   * If the command exits with a zero status code, that indicates the check is successful.
+   * If the command exits with a non-zero status code, the output of the command will be reported
+   * to provide details about the failure.
+   */
+  checkCommand: string;
+
+  /**
+   * Command to be ran to attempt to fix the check.
+   * If undefined, this check will not be fixable.
+   */
+  fixCommand?: string;
+}
+
+/**
+ * Create a check that runs a CLI command.
+ *
+ * @param options - options for creating the check, including the command to run, and, optionally, a command which can be run to attempt to fix the failing check.
+ * @returns - the created check
+ */
+export function scriptCheck(options: ScriptCheckOptions): Check {
+  return {
+    ...options,
+    hasFix: Boolean(options.fixCommand),
+    async check({ fix, project, verbose }) {
+      if (fix) {
+        check(options.fixCommand !== undefined, "fix command must be defined");
+        const { exitCode, output } = await run(options.fixCommand!, {
```

Comment:
you can avoid needing to put the `!` on `fixCommand` here if you annotate `check` as an assertion function: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions